src/reader.rs

```diff
@@ -80,3 +81,44 @@ impl<R: AsyncRead + AsyncSeek + Unpin + Send> Reader<R> {
         })
     }
 }
+
+/// Primary source used for reading required bytes for operations.
+#[allow(clippy::len_without_is_empty)]
+// TODO: async version
+pub trait ChunkReader {
+    type T: Read;
+
+    /// Get total length of bytes. Useful for parsing the metadata located at
+    /// the end of the file.
+    fn len(&self) -> u64;
+
+    /// Get a reader starting at a specific offset.
+    fn get_read(&self, offset_from_start: u64) -> std::io::Result<Self::T>;
+
+    /// Read bytes from an offset with specific length.
+    fn get_bytes(&self, offset_from_start: u64, length: u64) -> std::io::Result<Vec<u8>> {
+        let mut bytes = vec![0; length as usize];
+        self.get_read(offset_from_start)?
+            .take(length)
+            .read_exact(&mut bytes)?;
+        Ok(bytes)
+    }
+}
+
+impl ChunkReader for File {
+    type T = BufReader<File>;
+
+    fn len(&self) -> u64 {
+        self.metadata().map(|m| m.len()).unwrap_or(0u64)
+    }
+
+    /// Care needs to be taken when using this simulatenously as underlying
+    /// file descriptor is the same and will be affected by other invocations.
+    ///
+    /// See [`File::try_clone()`] for more details.
+    fn get_read(&self, offset_from_start: u64) -> std::io::Result<Self::T> {
+        let mut reader = self.try_clone()?;
+        reader.seek(SeekFrom::Start(offset_from_start))?;
+        Ok(BufReader::new(self.try_clone()?))
+    }
+}
```

Comment:
Essentially what is done in parquet at the moment. I like this interface as we don't have to keep doing the Seek ourselves